staging/src/k8s.io/dynamic-resource-allocation/cel/compile.go

```diff
@@ -322,3 +352,67 @@ func (m mapper) Find(key ref.Val) (ref.Val, bool) {
 
 	return m.defaultValue, true
 }
+
+// sizeEstimator tells the cost estimator the maximum size of maps or strings accessible through the `device` variable.
+// Without this, the maximum string size of e.g. `device.attributes["dra.example.com"].services` would be unknown.
+//
+// sizeEstimator is derived from the sizeEstimator in k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel.
+type sizeEstimator struct {
+	compiler *compiler
+}
+
+func (s *sizeEstimator) EstimateSize(element checker.AstNode) *checker.SizeEstimate {
```

Comment:
@liggitt for context, this is an exact copy of https://github.com/kubernetes/kubernetes/blob/3d2ee2fbb729857e5f71868107882f9e3c2aa43e/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/compilation.go#L313.  We have plans to follow up this PR and centralize the logic in a exportable way after this PR.  I mention it here to highlight that this is a well tested routine that we use in CRD validation rules.

Reply:
> @liggitt for context, this is an exact copy of https://github.com/kubernetes/kubernetes/blob/3d2ee2fbb729857e5f71868107882f9e3c2aa43e/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/compilation.go#L313

is it? these bits look unique to this impl, and I don't have context to know how that impacts the logic

```
	switch path[0] {
	case deviceVar:
		currentNode = s.compiler.deviceType
	default:
		// Unknown root, shouldn't happen.
		return nil
	}
```

```
			if !ok {
				// If this is an attribute map, then we know that all elements
				// have the same maximum size as set in attributeType, regardless
				// of their name.
				if currentNode.ElemType == attributeType {
					currentNode = attributeType
					continue
				}
				return nil
			}
```

Reply:
That's the "adaption to DRA" that I mentioned earlier.